addons/mail/models/mail_activity.py

```diff
@@ -572,53 +585,161 @@ def action_open_document(self):
+                                    deadline of the related activity
+                       }
+        """
+        user_tz = self.user_id.sudo().tz
+        Model = self.env[res_model]
+
+        # 1. Retrieve all ongoing and completed activities according to the parameters
+        activity_types = self.env['mail.activity.type'].search([('res_model', 'in', (res_model, False))],
+                                                               order='sequence, id')
+        fetch_done = fetch_done and activity_types.filtered('keep_done')
         activity_domain = [('res_model', '=', res_model)]
-        if domain or limit or offset:
-            res = self.env[res_model].search(domain or [], limit=limit, offset=offset)
-            activity_domain.append(('res_id', 'in', res.ids))
-        grouped_activities = self.env['mail.activity']._read_group(
+        is_filtered = domain or limit or offset
+        domain_res_ids = Model._search(domain or [], offset, limit) if is_filtered else []
+        if is_filtered:
+            activity_domain.append(('res_id', 'in', domain_res_ids))
+        all_activities = self.env['mail.activity'].with_context(active_test=not fetch_done).search_read(
             activity_domain,
-            ['res_id', 'activity_type_id'],
-            ['id:array_agg', 'date_deadline:min', '__count'])
-        # filter out unreadable records
-        if not domain:
-            res_ids = tuple(res_id for res_id, *_ in grouped_activities)
-            res_ids_set = set(self.env[res_model].search([('id', 'in', res_ids)])._ids)
-            grouped_activities = [a for a in grouped_activities if a[0] in res_ids_set]
+            ['activity_type_id', 'res_id', 'date_deadline', 'user_id', 'active'] +
+            (['attachment_ids'] if fetch_done else []),
+        )
+        all_ongoing_activities, all_completed_activities = partition(lambda a: a['active'], all_activities)
+
+        # 2. Get attachment of completed activities
+        if all_completed_activities:
+            all_attachment_ids = [
+                attachment_id
+                for act in all_completed_activities
+                for attachment_id in act['attachment_ids']
+            ]
+            attachments_by_id = {
+                a['id']: a
+                for a in self.env['ir.attachment'].search_read([['id', 'in', all_attachment_ids]],
+                                                               ['create_date', 'name'])
+            } if all_attachment_ids else {}
+        else:
+            attachments_by_id = {}
+
+        # 3. Group activities per records and activity type
+        grouped_completed_activities = dict(groupby(all_completed_activities,
+                                                    key=lambda a: (a['res_id'], a['activity_type_id'][0])))
+        grouped_ongoing_activities = dict(groupby(all_ongoing_activities,
+                                                  key=lambda a: (a['res_id'], a['activity_type_id'][0])))
```

Comment:
Example of simpler variable name that does not impact readong IMO
```suggestion
        grouped_completed = dict(
            groupby(completed_activities, key=lambda a: (a['res_id'], a['activity_type_id'][0]))
        )
        grouped_ongoing = dict(
            groupby(ongoing_activities, key=lambda a: (a['res_id'], a['activity_type_id'][0]))
        )